# Nội dung chính:
1. Đặt vấn đề
2. TF-IDF
3. Thực hành tự xây dựng mô hình TF-IDF
4. Consine Similarity
5. Các biến thể của TF-IDF
6. Sử dụng sklearn để build mô hình TF-IDF

Ở các bài viết trước, chúng ta đã tìm hiểu qua 2 mô hình word embbeding (nhúng từ) cơ bản nhất là one-hot endcoding và bag of word. Trong bài này chúng ta sẽ tìm hiểu thêm 1 mô hình nữa phổ biến hơn nhưng vẫn đơn giản đó là TF-IDF.

# Đặt vấn đề?

BoW là phương pháp đếm số từ. Vậy Bow có thực sự hiệu quả không?<br/>
Ví dụ ta có 2 văn bản, một văn bản có tổng số từ là 20 từ nhưng có đến 10 từ là "chó". Con văn bản còn lại có đến 10000 từ và có 2000 từ là "chó". Nếu theo phương pháp BoW thì rõ ràng văn bản thứ 2 là từ "chó" có tầm quan trọng hơn.<br/>
Tuy nhiên, không phải như vậy, với văn bản thứ nhất thì từ "chó" chiếm đến 50% số từ của văn bản. Còn văn bản thứ 2 thì chỉ chiếm 20%. Do vậy, "chó" ở văn bản thứ nhất có tầm ảnh hưởng lớn hơn so với "chó" của văn bản thứ 2. Từ đó, chúng ta thấy rằng, BoW chưa thể hiện đúng được tầm ảnh hưởng của 1 từ trong văn bản. Vì thế, TF-IDF ra đời và khắc phục điều này.

# TF-IDF

<b>tf–idf</b>, viết tắt của thuật ngữ tiếng Anh <i>term frequency – inverse document frequency</i>, của một từ là một con số thu được qua thống kê thể hiện mức độ quan trọng của từ này trong một văn bản, mà bản thân văn bản đang xét nằm trong một tập hợp các văn bản. Giá trị cao thể hiện độ quan trọng cao và nó phụ thuộc vào số lần từ xuất hiện trong văn bản nhưng bù lại bởi tần suất của từ đó trong tập dữ liệu. Một vài biến thể của tf-idf thường được sử dụng trong các hệ thống tìm kiếm như một công cụ chính để đánh giá và sắp xếp văn bản dựa vào truy vấn của người dùng.  Tf-idf cũng được sử dụng để lọc những từ stopwords trong các bài toán như tóm tắt văn bản và phân loại văn bản.

## TF là gì?

<b>TF</b>: Term Frequency (Tần suất xuất hiện của từ) là số lần từ xuất hiện trong văn bản. Term frequency thường được chia cho độ dài văn bản (Tổng số từ trong một văn bản)

$$tf(t,d) = \frac{f(t,d)}{n}$$

Trong đó:
    * tf(t, d): tần suất xuất hiện từ t trong văn bản d
    * f(t,d): Số lần xuất hiện của từ t trong văn bản d
    * n: Tổng số từ văn bản d

## IDF là gì?

Inverse Document Frequency(Nghịch đảo tần suất của văn bản), giúp đánh giá tầm quan trọng của một từ . Khi tính toán TF , tất cả các từ được coi như có độ quan trọng bằng nhau. Nhưng  một số từ như “is”, “of” và “that” thường xuất hiện rất nhiều lần nhưng độ quan trọng là không cao. Như thế chúng ta cần giảm độ quan trọng của những từ này xuống.

<img src="./images/idf.png"/>

Trong đó:
    * idf(t, D): giá trị idf của từ t trong tập văn bản.
    * |D|: Tổng số văn bản trong tập D.
    * |{d ∈ D : t ∈ d}|: thể hiện số văn bản trong tập D có chứa từ t.

Cơ số logarit trong công thức này không thay đổi giá trị idf của từ mà chỉ thu hẹp khoảng giá trị của từ đó. Vì thay đổi cơ số sẽ dẫn đến việc giá trị của các từ thay đổi bởi một số nhất định và tỷ lệ giữa các trọng lượng với nhau sẽ không thay đổi. (nói cách khác, thay đổi cơ số sẽ không ảnh hưởng đến tỷ lệ giữa các giá trị IDF). Việc sử dụng logarit nhằm giúp giá trị tf-idf của một từ nhỏ hơn, do chúng ta có công thức tính tf-idf của một từ trong 1 văn bản là tích của tf và idf của từ đó.

Cụ thể, chúng ta có công thức tính <b>tf-idf</b> hoàn chỉnh như sau: <font color="blue"><b>tfidf(t, d, D) = tf(t, d) x idf(t, D)</b></font>

Những từ có giá trị TF-IDF cao là những từ xuất hiện nhiều trong văn bản này, và xuất hiện ít trong các văn bản khác. Việc này giúp lọc ra những từ phổ biến và giữ lại những từ có giá trị cao (từ khoá của văn bản đó).

<b>Ví dụ:</b> một văn bản chưa 100 từ mà từ “cat” xuất hiện 3 lần . Term frequency cho từ cat này là (3/100) = 0.03 . Giả sử , chúng ta có 10 triệu văn bản và từ “cat” xuất hiện trong một nghìn văn bản . Như thế , idf được dính là  log(10,000,000 / 1,000) = 4. Như thế , tf-idf của từ “cat” trong văn bản này sẽ là : 0.03 * 4 = 0.12

# Thực hành xây dựng mô hình TF-IDF

Năm 1998, Google xử lý trung bình 9800 tìm kiếm mỗi ngày. Năm 2016, số lượng tìm kiếm trung bình mỗi ngày trên Google là hơn 9 tỉ lượt tìm kiếm. Biểu đồ dưới đây cho ta thông tin chi tiết về mức tăng tương về số lần tìm kiếm trên Google.

<img src="./images/Google.jpg"/>

Lý do chính trong thành công của Google chính là thuật toán PageRank. Thuật toán PageRank quyết định xem danh tiếng và độ tin cậy của một website như thế nào. Nhưng bên cạnh đó còn có một phần khác nữa đã tạo nên thành công của Google. Cụm từ tìm kiếm (query) do người dùng nhập vào cần được xử lý để matching với các tài liệu trên Internet. Bài viết sẽ tập trung vào phần thứ hai này. Tôi sẽ tạo ra ba document (tài liệu) để mô tả phần matching này hoạt động như thế nào.

In [1]:
doc1 = "Người lên ngựa kẻ chia bào. Rừng phong thu đã nhóm màu quan san"
doc2 = "Ô hay buồn vương ây ngô đồng. Vàng rơi vàng rơi thu mênh mông"
doc3 = "Một chiều vê bến sông thu. Nghe tin em cưới á cái đù"

Hãy tưởng tượng rằng chúng ta đang tìm kiếm trong một data set hoặc trên Internet ( và giả sử Internet chỉ có 3 tài liệu này) với từ khóa chúng ta nhập là: <b>sông thu</b>.

Tìm kiếm của chúng ta là một free text query, nghĩa là các từ khóa được gõ một cách tùy ý trong ô tìm kiếm và không có sự kết nối giữa các từ với nhau. Theo cách tìm kiếm này, nếu chúng ta nhập sông thu hay thu sông thì thuật toán cũng shows cho chúng ta các kết quả giống nhau.<br/>

Hãy đi vào chi tiết từng bước để xem phần matching tìm kiếm này hoạt động như thế nào.

## Bước 1: Tearm Frequency (TF - hay tần số xuất hiện của một từ)

In [2]:
# Tiền xử lý: loại bỏ dấu chấm và chuyển về chữ thường
doc1 = doc1.replace("."," ").lower()
doc2 = doc2.replace("."," ").lower()
doc3 = doc3.replace("."," ").lower()
print("Văn bản 1:", doc1)
print("Văn bản 2:", doc2)
print("Văn bản 3:", doc3)

Văn bản 1: người lên ngựa kẻ chia bào  rừng phong thu đã nhóm màu quan san
Văn bản 2: ô hay buồn vương ây ngô đồng  vàng rơi vàng rơi thu mênh mông
Văn bản 3: một chiều vê bến sông thu  nghe tin em cưới á cái đù


In [3]:
# Cắt lấy từ (tokenizer)
word1 = doc1.split()
word2 = doc2.split()
word3 = doc3.split()
print("Word1:", word1)
print("Word2:", word2)
print("Word3:", word3)

Word1: ['người', 'lên', 'ngựa', 'kẻ', 'chia', 'bào', 'rừng', 'phong', 'thu', 'đã', 'nhóm', 'màu', 'quan', 'san']
Word2: ['ô', 'hay', 'buồn', 'vương', 'ây', 'ngô', 'đồng', 'vàng', 'rơi', 'vàng', 'rơi', 'thu', 'mênh', 'mông']
Word3: ['một', 'chiều', 'vê', 'bến', 'sông', 'thu', 'nghe', 'tin', 'em', 'cưới', 'á', 'cái', 'đù']


In [4]:
# Đếm số lần xuất hiện
from collections import Counter
word_dict1 = Counter(word1)
word_dict2 = Counter(word2)
word_dict3 = Counter(word3)
print(word_dict1)
print(word_dict2)
print(word_dict3)

Counter({'người': 1, 'lên': 1, 'ngựa': 1, 'kẻ': 1, 'chia': 1, 'bào': 1, 'rừng': 1, 'phong': 1, 'thu': 1, 'đã': 1, 'nhóm': 1, 'màu': 1, 'quan': 1, 'san': 1})
Counter({'vàng': 2, 'rơi': 2, 'ô': 1, 'hay': 1, 'buồn': 1, 'vương': 1, 'ây': 1, 'ngô': 1, 'đồng': 1, 'thu': 1, 'mênh': 1, 'mông': 1})
Counter({'một': 1, 'chiều': 1, 'vê': 1, 'bến': 1, 'sông': 1, 'thu': 1, 'nghe': 1, 'tin': 1, 'em': 1, 'cưới': 1, 'á': 1, 'cái': 1, 'đù': 1})


In [5]:
# Tính toán TF
def computeTF(word_dict, words):
    tf_dict = {}
    words_count = len(words)
    for word, count in word_dict.items():
        tf_dict[word] = count/float(words_count)
    return tf_dict

In [6]:
tf_doc1 = computeTF(word_dict1, word1)
tf_doc2 = computeTF(word_dict2, word2)
tf_doc3 = computeTF(word_dict3, word3)

In [7]:
tf_doc1

{'người': 0.07142857142857142,
 'lên': 0.07142857142857142,
 'ngựa': 0.07142857142857142,
 'kẻ': 0.07142857142857142,
 'chia': 0.07142857142857142,
 'bào': 0.07142857142857142,
 'rừng': 0.07142857142857142,
 'phong': 0.07142857142857142,
 'thu': 0.07142857142857142,
 'đã': 0.07142857142857142,
 'nhóm': 0.07142857142857142,
 'màu': 0.07142857142857142,
 'quan': 0.07142857142857142,
 'san': 0.07142857142857142}

In [8]:
tf_doc2

{'ô': 0.07142857142857142,
 'hay': 0.07142857142857142,
 'buồn': 0.07142857142857142,
 'vương': 0.07142857142857142,
 'ây': 0.07142857142857142,
 'ngô': 0.07142857142857142,
 'đồng': 0.07142857142857142,
 'vàng': 0.14285714285714285,
 'rơi': 0.14285714285714285,
 'thu': 0.07142857142857142,
 'mênh': 0.07142857142857142,
 'mông': 0.07142857142857142}

In [9]:
tf_doc3

{'một': 0.07692307692307693,
 'chiều': 0.07692307692307693,
 'vê': 0.07692307692307693,
 'bến': 0.07692307692307693,
 'sông': 0.07692307692307693,
 'thu': 0.07692307692307693,
 'nghe': 0.07692307692307693,
 'tin': 0.07692307692307693,
 'em': 0.07692307692307693,
 'cưới': 0.07692307692307693,
 'á': 0.07692307692307693,
 'cái': 0.07692307692307693,
 'đù': 0.07692307692307693}

## Bước 2: Inverse Document Frequency (IDF – tần số nghịch của một từ trong một data set)

Mục đích chính của việc tìm kiếm văn bản là tìm ra những văn bản trong một data set (hoặc Internet nói chung) có nội dung liên quan nhất với từ tìm kiếm của người dùng. Ở bước 1 (TF), tất cả các từ đều được đánh giá quan trọng ngang nhau.Tuy nhiên, trong thực tế thì có một số từ xuất hiện quá nhiều và không có vai trò quyết định trong việc tìm ra các văn bản có nội dung liên quan với chủ đề mà người dùng tìm kiếm (trong tiếng Anh thì những từ xuất hiện nhiều và ít quan trọng là the, a, he, she, etc…). Chúng ta cần tìm ra một cách để làm giảm trọng số của những từ xuất hiện quá thường xuyên (trên Internet) và tăng trọng số của những từ ít xuất hiện (trên Internet) hơn. Và công cụ toán học Logarithm (Lôgarit) sẽ giúp chúng ta làm được điều này.

In [10]:
# Tính toán IDF
import math
def computeIDF(word_dicts):
    idf_dict = {}
    N = len(word_dicts)
    keys = set().union(*word_dicts)
    
    # Tạo danh sách keys
    idf_dict = dict.fromkeys(keys, 0)
    
    for doc in word_dicts: # duyệt từng document
        for word, val in doc.items(): # duyệt từng từ trong document
            if val > 0:
                idf_dict[word] += 1
                
    for word, val in idf_dict.items():
        idf_dict[word] = math.log(N / float(val))
        
    return idf_dict

In [11]:
idfs = computeIDF([word_dict1, word_dict2, word_dict3])

In [12]:
idfs

{'quan': 1.0986122886681098,
 'bào': 1.0986122886681098,
 'hay': 1.0986122886681098,
 'vê': 1.0986122886681098,
 'một': 1.0986122886681098,
 'mênh': 1.0986122886681098,
 'san': 1.0986122886681098,
 'tin': 1.0986122886681098,
 'sông': 1.0986122886681098,
 'ngựa': 1.0986122886681098,
 'ngô': 1.0986122886681098,
 'ây': 1.0986122886681098,
 'á': 1.0986122886681098,
 'cái': 1.0986122886681098,
 'thu': 0.0,
 'đồng': 1.0986122886681098,
 'cưới': 1.0986122886681098,
 'đù': 1.0986122886681098,
 'buồn': 1.0986122886681098,
 'người': 1.0986122886681098,
 'chia': 1.0986122886681098,
 'chiều': 1.0986122886681098,
 'đã': 1.0986122886681098,
 'rơi': 1.0986122886681098,
 'lên': 1.0986122886681098,
 'màu': 1.0986122886681098,
 'vương': 1.0986122886681098,
 'phong': 1.0986122886681098,
 'nhóm': 1.0986122886681098,
 'vàng': 1.0986122886681098,
 'nghe': 1.0986122886681098,
 'rừng': 1.0986122886681098,
 'bến': 1.0986122886681098,
 'mông': 1.0986122886681098,
 'em': 1.0986122886681098,
 'ô': 1.0986122886681

Bên dưới là bảng IDF của tất cả các từ trong data set. Trong đó từ <b>thu</b> xuất hiện trong cả 3 văn bản nên nó sẽ có điểm số IDF thấp hơn so với các từ chỉ xuất hiện một văn bản.

## Bước 3: Tính TF*IDF

In [13]:
def computeTFIDF(tf, idf):
    tf_idf = {}
    for word, val in tf.items():
        tf_idf[word] = val*idf[word]
    return tf_idf

In [14]:
tf_idf_doc1 = computeTFIDF(tf_doc1, idfs)
tf_idf_doc2 = computeTFIDF(tf_doc2, idfs)
tf_idf_doc3 = computeTFIDF(tf_doc3, idfs)

In [15]:
tf_idf_doc1

{'người': 0.07847230633343641,
 'lên': 0.07847230633343641,
 'ngựa': 0.07847230633343641,
 'kẻ': 0.07847230633343641,
 'chia': 0.07847230633343641,
 'bào': 0.07847230633343641,
 'rừng': 0.07847230633343641,
 'phong': 0.07847230633343641,
 'thu': 0.0,
 'đã': 0.07847230633343641,
 'nhóm': 0.07847230633343641,
 'màu': 0.07847230633343641,
 'quan': 0.07847230633343641,
 'san': 0.07847230633343641}

In [16]:
tf_idf_doc2

{'ô': 0.07847230633343641,
 'hay': 0.07847230633343641,
 'buồn': 0.07847230633343641,
 'vương': 0.07847230633343641,
 'ây': 0.07847230633343641,
 'ngô': 0.07847230633343641,
 'đồng': 0.07847230633343641,
 'vàng': 0.15694461266687282,
 'rơi': 0.15694461266687282,
 'thu': 0.0,
 'mênh': 0.07847230633343641,
 'mông': 0.07847230633343641}

In [17]:
tf_idf_doc3 

{'một': 0.08450863758985461,
 'chiều': 0.08450863758985461,
 'vê': 0.08450863758985461,
 'bến': 0.08450863758985461,
 'sông': 0.08450863758985461,
 'thu': 0.0,
 'nghe': 0.08450863758985461,
 'tin': 0.08450863758985461,
 'em': 0.08450863758985461,
 'cưới': 0.08450863758985461,
 'á': 0.08450863758985461,
 'cái': 0.08450863758985461,
 'đù': 0.08450863758985461}

In [18]:
import pandas as pd
data = pd.DataFrame([tf_idf_doc1, tf_idf_doc2, tf_idf_doc3]).T

In [19]:
data

0         1         2
người  0.078472       NaN       NaN
lên    0.078472       NaN       NaN
ngựa   0.078472       NaN       NaN
kẻ     0.078472       NaN       NaN
chia   0.078472       NaN       NaN
bào    0.078472       NaN       NaN
rừng   0.078472       NaN       NaN
phong  0.078472       NaN       NaN
thu    0.000000  0.000000  0.000000
đã     0.078472       NaN       NaN
nhóm   0.078472       NaN       NaN
màu    0.078472       NaN       NaN
quan   0.078472       NaN       NaN
san    0.078472       NaN       NaN
ô           NaN  0.078472       NaN
hay         NaN  0.078472       NaN
buồn        NaN  0.078472       NaN
vương       NaN  0.078472       NaN
ây          NaN  0.078472       NaN
ngô         NaN  0.078472       NaN
đồng        NaN  0.078472       NaN
vàng        NaN  0.156945       NaN
rơi         NaN  0.156945       NaN
mênh        NaN  0.078472       NaN
mông        NaN  0.078472       NaN
một         NaN       NaN  0.084509
chiều       NaN       NaN  0.084509
vê          NaN       NaN  0.084509
bến         NaN       NaN  0.084509
sông        NaN       NaN  0.084509
nghe        NaN       NaN  0.084509
tin         NaN       NaN  0.084509
em          NaN       NaN  0.084509
cưới        NaN       NaN  0.084509
á           NaN       NaN  0.084509
cái         NaN       NaN  0.084509
đù          NaN       NaN  0.084509

Hãy nhớ rằng mục đích của chúng ta là đi tìm văn bản có nội dung liên quan nhất cho cụm từ tìm kiếm : <b>sông thu</b>.

In [20]:
data.loc["sông"]

0         NaN
1         NaN
2    0.084509
Name: sông, dtype: float64

In [21]:
data.loc["thu"]

0    0.0
1    0.0
2    0.0
Name: thu, dtype: float64

Vậy tính xong TF-IDF, ta cần phải có 1 công thức để đo mức độ tương quan. Vì vậy chúng ta cung tìm hiểu phần sau:

# Consine Similarity

Consine simalarity là 1 công thức đo mức độ tương đồng giữa 2 vector trong một không gian.

$$similarity=cos(\theta)=\frac{A.B}{||A||.||B||}$$

Chúng ta mô tả mỗi document như là một vector. Một data set được xem như là một tập hợp các vector trong một không gian vector. Mỗi từ trong không gian vector sẽ có trục của riêng nó. Bằng cách sử dụng công thức phía trên, chúng ta có thể tìm ra độ tương đồng của bất kì tài liệu nào.

Vector chỉ làm việc với con số. Trong bài viết này chúng ta đang làm việc với văn bản. Đó là lý do tại sao chúng ta sử dụng TF – IDF để chuyển đổi từ ngữ văn bản thành số để có thể biểu diễn chúng ở dạng vector.

Cụm từ tìm kiếm của người dùng cũng được xem là một vector. Chúng ta tính giá trị TF * IDF cho cụm từ truy vấn.

In [22]:
tf_idf_user = {}
tf_idf_user["sông"] = 0.5*idfs["sông"]
tf_idf_user["thu"] = 0.5*idfs["thu"]
tf_idf_user

{'sông': 0.5493061443340549, 'thu': 0.0}

0.5 ở đây tức là cụm từ tìm kiếm có 2 từ "sông thu" nên tf_user["sông"] = tf_user["thu"] = 0.5

Bây giờ chúng ta hãy tính cosine similarity (tương đồng cosine) giữa cụm từ tìm kiếm (Query) và các Document.

In [23]:
import math
def consine_sim(vec1, vec2):
    """ Let's convert our dictionaries to lists for easier matching """
    vec1 = [val for val in vec1.values()]
    vec2 = [val for val in vec2.values()]
    
    dot_prod = 0
    for i,v in enumerate(vec1):
        dot_prod += v*vec2[i]
    
    mag_1 = math.sqrt(sum([x**2 for x in vec1]))
    mag_2 = math.sqrt(sum([x**2 for x in vec2]))
    
    return dot_prod/(mag_1*mag_2)

In [24]:
vec1 = tf_idf_user
vec2 = {}
vec2["sông"] = 0.0 # do trong tf_idf_doc1 không có từ sông
vec2["thu"] = tf_idf_doc1["thu"]
print("vec1:",vec1)
print("vec2:",vec2)

vec1: {'sông': 0.5493061443340549, 'thu': 0.0}
vec2: {'sông': 0.0, 'thu': 0.0}


Bây giờ tính sự tương đồng giữa từ khoá người dùng và document1 thành bài toán trở thành tính độ tương đồng giữa vec1 = (0.549, 0) và vec2=(0,0)

In [25]:
try:
    consine_sim(vec1, vec2)
except:
    print("Lỗi chia cho 0")

Lỗi chia cho 0


Lỗi. Ta thử tính với document3 xem sao.

In [26]:
vec2_3 = {}
vec2_3["sông"] = tf_idf_doc3["sông"]
vec2_3["thu"] = tf_idf_doc3["thu"]
print("vec1:",vec1)
print("vec2:",vec2_3)

vec1: {'sông': 0.5493061443340549, 'thu': 0.0}
vec2: {'sông': 0.08450863758985461, 'thu': 0.0}


In [27]:
consine_sim(vec1, vec2_3)

1.0

Vậy là có cái tính được cái không? Một số bị lỗi chia cho 0. Và để khắc phục điều này người ta có một số phương pháp cải tiến TF-IDF.

# Các biến thể của TF-IDF

Với bài toán trên, ta thấy từ thu xuất hiện ở cả 3 văn bản và idf["thu"]=0 và gây ra hiện tượng ta không thể tính được consine similarity. Vì vậy ý tưởng cơ bản nhất là ta điều chỉnh công thức tính idf.

IDF(thu) = 1 + ln(Tổng số văn bản trong data set/Số văn bản chứa từ thu)

In [28]:
# Tính toán IDF
import math
def computeIDF_new(word_dicts):
    idf_dict = {}
    N = len(word_dicts)
    keys = set().union(*word_dicts)
    
    # Tạo danh sách keys
    idf_dict = dict.fromkeys(keys, 0)
    
    for doc in word_dicts: # duyệt từng document
        for word, val in doc.items(): # duyệt từng từ trong document
            if val > 0:
                idf_dict[word] += 1
                
    for word, val in idf_dict.items():
        idf_dict[word] = 1 + math.log(N / float(val)) # Thay đổi tại đây
        
    return idf_dict

In [29]:
idfs_new = computeIDF_new([word_dict1, word_dict2, word_dict3])

In [30]:
idfs_new

{'quan': 2.09861228866811,
 'bào': 2.09861228866811,
 'hay': 2.09861228866811,
 'vê': 2.09861228866811,
 'một': 2.09861228866811,
 'mênh': 2.09861228866811,
 'san': 2.09861228866811,
 'tin': 2.09861228866811,
 'sông': 2.09861228866811,
 'ngựa': 2.09861228866811,
 'ngô': 2.09861228866811,
 'ây': 2.09861228866811,
 'á': 2.09861228866811,
 'cái': 2.09861228866811,
 'thu': 1.0,
 'đồng': 2.09861228866811,
 'cưới': 2.09861228866811,
 'đù': 2.09861228866811,
 'buồn': 2.09861228866811,
 'người': 2.09861228866811,
 'chia': 2.09861228866811,
 'chiều': 2.09861228866811,
 'đã': 2.09861228866811,
 'rơi': 2.09861228866811,
 'lên': 2.09861228866811,
 'màu': 2.09861228866811,
 'vương': 2.09861228866811,
 'phong': 2.09861228866811,
 'nhóm': 2.09861228866811,
 'vàng': 2.09861228866811,
 'nghe': 2.09861228866811,
 'rừng': 2.09861228866811,
 'bến': 2.09861228866811,
 'mông': 2.09861228866811,
 'em': 2.09861228866811,
 'ô': 2.09861228866811,
 'kẻ': 2.09861228866811}

In [31]:
tf_idf_doc1_new = computeTFIDF(tf_doc1, idfs_new)
tf_idf_doc2_new = computeTFIDF(tf_doc2, idfs_new)
tf_idf_doc3_new = computeTFIDF(tf_doc3, idfs_new)

In [32]:
tf_idf_doc1_new

{'người': 0.14990087776200786,
 'lên': 0.14990087776200786,
 'ngựa': 0.14990087776200786,
 'kẻ': 0.14990087776200786,
 'chia': 0.14990087776200786,
 'bào': 0.14990087776200786,
 'rừng': 0.14990087776200786,
 'phong': 0.14990087776200786,
 'thu': 0.07142857142857142,
 'đã': 0.14990087776200786,
 'nhóm': 0.14990087776200786,
 'màu': 0.14990087776200786,
 'quan': 0.14990087776200786,
 'san': 0.14990087776200786}

In [33]:
tf_idf_user_new = {}
tf_idf_user_new["sông"] = 0.5*idfs_new["sông"]
tf_idf_user_new["thu"] = 0.5*idfs_new["thu"]
tf_idf_user_new

{'sông': 1.049306144334055, 'thu': 0.5}

In [34]:
vec_doc1_new = {}
vec_doc1_new["sông"] = 0.0
vec_doc1_new["thu"] = tf_idf_doc1_new["thu"]
vec_doc1_new

{'sông': 0.0, 'thu': 0.07142857142857142}

In [35]:
vec_doc2_new = {}
vec_doc2_new["sông"] = 0.0
vec_doc2_new["thu"] = tf_idf_doc2_new["thu"]
vec_doc2_new

{'sông': 0.0, 'thu': 0.07142857142857142}

In [36]:
vec_doc3_new = {}
vec_doc3_new["sông"] = tf_idf_doc3_new["sông"]
vec_doc3_new["thu"] = tf_idf_doc3_new["thu"]
vec_doc3_new

{'sông': 0.16143171451293153, 'thu': 0.07692307692307693}

In [37]:
consine_sim(tf_idf_user_new, vec_doc1_new)

0.430165282498796

In [38]:
consine_sim(tf_idf_user_new, vec_doc2_new)

0.430165282498796

In [39]:
consine_sim(tf_idf_user_new, vec_doc3_new)

1.0

Chúng ta có thể thấy là Document 3 có score cao nhất bằng 1. Lý do là vì Document 3 chứa đồng thời cả hai từ sông và thu. Điều này có nghĩa là khi người dùng tìm kiếm với cụm từ là sông thu thì thuật toán sẽ trả về tài liệu có độ tương đồng cosine lớn nhất với nó là Document 3. Tất nhiên, mô hình trên là tôi xây dựng tay nên chưa tối ưu, khi làm bài toán thực tế, chúng ta sẽ thường sử dụng thư viện hỗ trợ.

Ngoài ra chúng ta có bảng sau:

<img src="./images/table.png"/>

# Sử dụng sklearn để build mô hình TF-IDF

Bây giờ có rất nhiều packet hỗ trợ chúng ta tự động hoá. Như là scikit-learn.Tại đây, bạn có thể sử dụng sklearn để xây dựng một ma trận TF-IDF. 

In [40]:
corpus = [doc1, doc2, doc3]
corpus

['người lên ngựa kẻ chia bào  rừng phong thu đã nhóm màu quan san',
 'ô hay buồn vương ây ngô đồng  vàng rơi vàng rơi thu mênh mông',
 'một chiều vê bến sông thu  nghe tin em cưới á cái đù']

In [41]:
lexicon = sorted(set(word1+word2+word3))
lexicon

['buồn',
 'bào',
 'bến',
 'chia',
 'chiều',
 'cái',
 'cưới',
 'em',
 'hay',
 'kẻ',
 'lên',
 'màu',
 'mênh',
 'mông',
 'một',
 'nghe',
 'ngô',
 'người',
 'ngựa',
 'nhóm',
 'phong',
 'quan',
 'rơi',
 'rừng',
 'san',
 'sông',
 'thu',
 'tin',
 'vàng',
 'vê',
 'vương',
 'á',
 'ây',
 'ô',
 'đã',
 'đù',
 'đồng']

In [42]:
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer = TfidfVectorizer(smooth_idf=False) # Mặc định là True
# Khi smooth_idf=False thì idf(t) = log [ n / df(t) ] + 1
# Còn True thì idf(t) = log [ n / (df(t) + 1) ]).
model = vectorizer.fit_transform(corpus)
print(vectorizer.get_feature_names())
print(model.todense().round(3))

['buồn', 'bào', 'bến', 'chia', 'chiều', 'cái', 'cưới', 'em', 'hay', 'kẻ', 'lên', 'màu', 'mênh', 'mông', 'một', 'nghe', 'ngô', 'người', 'ngựa', 'nhóm', 'phong', 'quan', 'rơi', 'rừng', 'san', 'sông', 'thu', 'tin', 'vàng', 'vê', 'vương', 'ây', 'đã', 'đù', 'đồng']
[[0.    0.275 0.    0.275 0.    0.    0.    0.    0.    0.275 0.275 0.275
  0.    0.    0.    0.    0.    0.275 0.275 0.275 0.275 0.275 0.    0.275
  0.275 0.    0.131 0.    0.    0.    0.    0.    0.275 0.    0.   ]
 [0.248 0.    0.    0.    0.    0.    0.    0.    0.248 0.    0.    0.
  0.248 0.248 0.    0.    0.248 0.    0.    0.    0.    0.    0.496 0.
  0.    0.    0.118 0.    0.496 0.    0.248 0.248 0.    0.    0.248]
 [0.    0.    0.298 0.    0.298 0.298 0.298 0.298 0.    0.    0.    0.
  0.    0.    0.298 0.298 0.    0.    0.    0.    0.    0.    0.    0.
  0.    0.298 0.142 0.298 0.    0.298 0.    0.    0.    0.298 0.   ]]


Mô hình của sklearn loại bỏ từ <b>á</b> và <b>ô</b> khỏi lexicon của chúng ta vì thư viện này cho rằng các từ 1 ký tự không đem lại ý nghĩa. Vì vậy kết quả tính mới hơi khác mô hình tự thiết lập của chúng ta, các bạn có thể thử loại bỏ 2 từ trên và tính toán lại để kiếm tra lại kết quả tự tính và của sklearn.

Xem thêm: https://github.com/scikit-learn/scikit-learn/blob/7a636f0cc5f63aaab2d0e556fec46989465934aa/sklearn/feature_extraction/text.py#L1140